### DISCUSSION TITLE
ScreenTrace API on Android is missing?

### DISCUSSION THREAD

**Discussion by shamilovtim** [invertase/react-native-firebase#6247] [General]
Android has an API for custom ScreenTraces: https://firebase.google.com/codelabs/measure-android-view-performance?hl=en#7. I noticed that react-native-firebase only has the regular (duration) trace API but not the ScreenTrace API. Would it be possible to add screentrace?

**Comment by mikehardy** [comment]
Hmmm :thinking: that's an interesting idea

It's important to note that this is not an official API, this is a re-use of official APIs, in native code, to extend existing tracing to fragments. The existing screen trace functionality already automatically (unless disabled by config) collects this information, it's just at a different granularity (Activity, vs Fragment). This is a codelab (read as: programming exploration) to extend to fragment granularity.

Does this even have applicability in a react-native app? I'm not sure it does as I'm not sure react-native uses fragments much if at all? I am not well informed in this area though, so that's my first question: does react-native use fragments at all? If not then there would be no point implementing this.

If it does, then this *could* be useful. If you wanted to add it, you would need to add the custom ScreenTrace extension code and figure out how to install the fragment manager lifecycle callbacks in there somehow, in such a way that at best everyone just got it "for free" (no native code changes required) or at worst it was just a one-liner to add in MainActivity for those that wanted it. Ideally it would have a config param in firebase.json that turned it off though defaulting to on is probably what the majority would want

This is not an official firebase API so I won't have time myself to directly implement it but I would collaborate with you on a PR if you wanted to contribute the code

**Comment by shamilovtim** [comment]
> Hmmm :thinking: that's an interesting idea
> 
> 
> 
> It's important to note that this is not an official API, this is a re-use of official APIs, in native code, to extend existing tracing to fragments. The existing screen trace functionality already automatically (unless disabled by config) collects this information, it's just at a different granularity (Activity, vs Fragment). This is a codelab (read as: programming exploration) to extend to fragment granularity.
> 
> 
> 
> Does this even have applicability in a react-native app? I'm not sure it does as I'm not sure react-native uses fragments much if at all? I am not well informed in this area though, so that's my first question: does react-native use fragments at all? If not then there would be no point implementing this.
> 
> 
> 
> If it does, then this *could* be useful. If you wanted to add it, you would need to add the custom ScreenTrace extension code and figure out how to install the fragment manager lifecycle callbacks in there somehow, in such a way that at best everyone just got it "for free" (no native code changes required) or at worst it was just a one-liner to add in MainActivity for those that wanted it. Ideally it would have a config param in firebase.json that turned it off though defaulting to on is probably what the majority would want
> 
> 
> 
> This is not an official firebase API so I won't have time myself to directly implement it but I would collaborate with you on a PR if you wanted to contribute the code

Thank you! I have a question on architecture. Based on what I see there, to use it for fragments they just instantiate a screen trace and call .start on it. What's to stop us from just adding .start to the bridged methods and being able to use it in a useeffect, and then .end to end the custom trace? We can pass the string for the name across the bridge. Do we really actually need to try to instrument fragments or could we just use this ad hoc across the bridge? I noticed the screen trace needed the {this} as an input (context? fragment?), would that be our blocker?

**Reply by shamilovtim** [reply]
Answering my own question looked at the custom class and it's just the activity. Looks easy enough to implement. I wonder if this should actually be part of react native firebase or should be a small outside package?

**Reply by mikehardy** [reply]
You could certainly start (and maybe always stay) as a separate small library, yes. The idea of it being ad-hoc sounds more interesting to me in the react-native context given my doubt on even native fragments mapping cleanly to what is a "react-native screen"

**Comment by mikehardy** [comment]
This may be interesting: https://firebase.google.com/support/release-notes/android#performance_v20-1-0
I'm planning on releasing react-native-firebase v14.12.0 today which will include the new firebase-android-sdk 30.1.0 that has this

**Reply by shamilovtim** [reply]
Awesome, thank you @mikehardy. I think I'd still need a screenTrace method that's usable from the Javascript side since we aren't split up into fragments (we don't use RN Native Stack). Is that on the roadmap or do you think I should do it?

**Reply by mikehardy** [reply]
Not on the roadmap at all sorry :sweat_smile: - current roadmap is: adopt firebase-ios-sdk v9 features, get jet-next (e2e testing harness) working well enough, port current e2e suite to jet-next, then get firebase-js-sdk v9 (modular style) syntax supported. Those are all repo-wide items I need to handle before diving into specific APIs. If you see me do any specific APIs between now and then it's just because I was "in the area" while triaging something else and it was trivial to test whatever the thing was so I could do it quickly

**Reply by shamilovtim** [reply]
Is there somewhere I could find the source code of this off the shelf screen trace? I'm a little confused about the verbiage: "Added support for out-of-the-box measurement of screen performance metrics." Is the implication that this needs to be called somewhere in the code or is the library going to do this by default without passing any kind of arguments to it?

**Reply by mikehardy** [reply]
Looks like it's been an epic feature, you can follow the PRs here and use those to surf in to the repo in its current state to see the implementation; https://github.com/firebase/firebase-android-sdk/pulls?q=is%3Apr+is%3Aclosed+FirePerf
